youtube_dl/downloader/external.py

```diff
@@ -192,15 +195,19 @@ def supports(cls, info_dict):
     def available(cls):
         return FFmpegPostProcessor().available
 
+    def get_execname(self):
+        return self.execname
```

Comment:
> It should instantiate FFmpegPostProcessor on it's own like available do

But that means the postprocessor gets initialized 3 times, and so it calls `_determine_executables()` 3 times. So it searches the system path 3 times to decide which program to run. In addition to being inefficient, it also means that if the set of installed programs changes while youtube-dl is running (unlikely), then it would report the wrong program.

```
pb3:youtube-dl jhawk$ python  -c 'import pdb, youtube_dl; pdb.runcall(youtube_dl.main, ["'$u'"])' youtube_dl
> /Users/jhawk/src/youtube-dl/youtube_dl/__init__.py(463)main()
-> try:
(Pdb) b youtube_dl/postprocessor/ffmpeg.py:77
Breakpoint 1 at /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py:77
(Pdb) commands 1
(com) bt
(com) continue
(Pdb) r
[generic] canadas-fast-track-refugee-plan-unanswered-questions-and-implications-for-us-national-security: Requesting header
WARNING: Falling back on generic information extractor.
[generic] canadas-fast-track-refugee-plan-unanswered-questions-and-implications-for-us-national-security: Downloading webpage
[generic] canadas-fast-track-refugee-plan-unanswered-questions-and-implications-for-us-national-security: Extracting information
[SenateISVP] govtaff020316: Downloading webpage
[SenateISVP] govtaff020316: Downloading f4m manifest
[SenateISVP] govtaff020316: Downloading m3u8 information
  /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/bdb.py(436)runcall()
-> res = func(*args, **kwds)
  /Users/jhawk/src/youtube-dl/youtube_dl/__init__.py(464)main()
-> _real_main(argv)
  /Users/jhawk/src/youtube-dl/youtube_dl/__init__.py(454)_real_main()
-> retcode = ydl.download(all_urls)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1884)download()
-> url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(772)extract_info()
-> return self.process_ie_result(ie_result, download, extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(833)process_ie_result()
-> extra_info=extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(772)extract_info()
-> return self.process_ie_result(ie_result, download, extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(826)process_ie_result()
-> return self.process_video_result(ie_result, download=download)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1483)process_video_result()
-> merger = FFmpegMergerPP(self)
  /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(58)__init__()
-> self._determine_executables()
> /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(77)_determine_executables()
-> programs = ['avprobe', 'avconv', 'ffmpeg', 'ffprobe']
> /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(77)_determine_executables()
-> programs = ['avprobe', 'avconv', 'ffmpeg', 'ffprobe']
[download] Destination: Integrated Senate Video Player-govtaff020316.mp4
  /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/bdb.py(436)runcall()
-> res = func(*args, **kwds)
  /Users/jhawk/src/youtube-dl/youtube_dl/__init__.py(464)main()
-> _real_main(argv)
  /Users/jhawk/src/youtube-dl/youtube_dl/__init__.py(454)_real_main()
-> retcode = ydl.download(all_urls)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1884)download()
-> url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(772)extract_info()
-> return self.process_ie_result(ie_result, download, extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(833)process_ie_result()
-> extra_info=extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(772)extract_info()
-> return self.process_ie_result(ie_result, download, extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(826)process_ie_result()
-> return self.process_video_result(ie_result, download=download)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1527)process_video_result()
-> self.process_info(new_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1791)process_info()
-> success = dl(filename, info_dict)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1733)dl()
-> return fd.download(name, info)
  /Users/jhawk/src/youtube-dl/youtube_dl/downloader/common.py(356)download()
-> return self.real_download(filename, info_dict)
  /Users/jhawk/src/youtube-dl/youtube_dl/downloader/external.py(32)real_download()
-> retval = self._call_downloader(tmpfilename, info_dict)
  /Users/jhawk/src/youtube-dl/youtube_dl/downloader/external.py(205)_call_downloader()
-> ffpp = FFmpegPostProcessor(downloader=self)
  /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(58)__init__()
-> self._determine_executables()
> /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(77)_determine_executables()
-> programs = ['avprobe', 'avconv', 'ffmpeg', 'ffprobe']
> /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(77)_determine_executables()
-> programs = ['avprobe', 'avconv', 'ffmpeg', 'ffprobe']


  /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/bdb.py(436)runcall()
-> res = func(*args, **kwds)
  /Users/jhawk/src/youtube-dl/youtube_dl/__init__.py(464)main()
-> _real_main(argv)
  /Users/jhawk/src/youtube-dl/youtube_dl/__init__.py(454)_real_main()
-> retcode = ydl.download(all_urls)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1884)download()
-> url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(772)extract_info()
-> return self.process_ie_result(ie_result, download, extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(833)process_ie_result()
-> extra_info=extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(772)extract_info()
-> return self.process_ie_result(ie_result, download, extra_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(826)process_ie_result()
-> return self.process_video_result(ie_result, download=download)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1527)process_video_result()
-> self.process_info(new_info)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1791)process_info()
-> success = dl(filename, info_dict)
  /Users/jhawk/src/youtube-dl/youtube_dl/YoutubeDL.py(1733)dl()
-> return fd.download(name, info)
  /Users/jhawk/src/youtube-dl/youtube_dl/downloader/common.py(356)download()
-> return self.real_download(filename, info_dict)
  /Users/jhawk/src/youtube-dl/youtube_dl/downloader/external.py(47)real_download()
-> self.get_execname(), retval))
  /Users/jhawk/src/youtube-dl/youtube_dl/downloader/external.py(201)get_execname()
-> return FFmpegPostProcessor().executable
  /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(58)__init__()
-> self._determine_executables()
> /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(77)_determine_executables()
-> programs = ['avprobe', 'avconv', 'ffmpeg', 'ffprobe']
> /Users/jhawk/src/youtube-dl/youtube_dl/postprocessor/ffmpeg.py(77)_determine_executables()
-> programs = ['avprobe', 'avconv', 'ffmpeg', 'ffprobe']
ERROR: avconv exited with code 1
```